README.md
# Multiscale Deformetrica

This code is an extension of the software Deformetrica.
It implements a coarse-to-fine strategy for atlas construction / image registration.

<div align="center"><img src="/digit_ctf_cv_k3_fold_4_colors.gif" width="300"/></div>

## Installation
      
    git clone https://github.com/fleurgaudfernau/Deformetrica_coarse_to_fine

    virtualenv -p python3 deformetrica && source deformetrica/bin/activate

    python3 -m pip install -r Deformetrica_multiscale/requirements.txt

    python3 -m pip install Deformetrica_multiscale/.
    
<hr style="border:1px solid #2b6777"/>

    virtualenv -p python3 deformetrica && source deformetrica/bin/activate
    
    pip install git+https://github.com/fleurgaudfernau/Deformetrica_multiscale.git
## About Deformetrica

Website: [www.deformetrica.org](http://www.deformetrica.org/)

**Deformetrica** is a software for the statistical analysis of 2D and 3D shape data. It essentially computes deformations of the 2D or 3D ambient space, which, in turn, warp any object embedded in this space, whether this object is a curve, a surface, a structured or unstructured set of points, an image, or any combination of them.

_Deformetrica_ comes with three main applications:
- **registration** : estimates the best possible deformation between two sets of objects;
- **atlas construction** : estimates an average object configuration from a collection of object sets, and the deformations from this average to each sample in the collection;
- **geodesic regression** : estimates an object time-series constrained to match as closely as possible a set of observations indexed by time.

_Deformetrica_ has very little requirements about the data it can deal with. In particular, it does __not__ require point correspondence between objects!

## References

Deformetrica relies on a control-points-based instance of the Large Deformation Diffeomorphic Metric Mapping framework, introduced in [\[Durrleman et al. 2014\]](https://linkinghub.elsevier.com/retrieve/pii/S1053811914005205). Are fully described in this article the **shooting**, **registration**, and **deterministic atlas** applications. Equipped with those fundamental building blocks, additional applications have been successively developed:
- the bayesian atlas, described in [\[Gori et al. 2017\]](https://hal.archives-ouvertes.fr/hal-01359423/);
- the geodesic regression, described in [\[Fishbaugh et al. 2017\]](https://www.medicalimageanalysisjournal.com/article/S1361-8415(17)30044-0/fulltext);
- the parallel transport, described in [\[Louis et al. 2018\]](https://www.researchgate.net/publication/319136479_Parallel_transport_in_shape_analysis_a_scalable_numerical_scheme);
- the longitudinal atlas, described in [\[Bône et al. 2018a\]](https://www.researchgate.net/publication/324037371_Learning_distributions_of_shape_trajectories_from_longitudinal_datasets_a_hierarchical_model_on_a_manifold_of_diffeomorphisms) and [\[Bône et al. 2020\]](https://www.researchgate.net/publication/342642363_Learning_the_spatiotemporal_variability_in_longitudinal_shape_data_sets).

[\[Bône et al. 2018b\]](https://www.researchgate.net/publication/327652245_Deformetrica_4_an_open-source_software_for_statistical_shape_analysis) provides a concise reference summarizing those functionalities, with unified notations.

# Archived repositories

- Deformetrica 3: [deformetrica-legacy2](https://gitlab.icm-institute.org/aramislab/deformetrica-legacy2)
- Deformetrica 2.1: [deformetrica-legacy](https://gitlab.icm-institute.org/aramislab/deformetrica-legacy)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## [4.3.0] - 2020-08-20
- Updates:
    - pytorch to 1.6
    - pykeops to 1.4.1



## [4.2.0] - 2019-04-15
=======
- Improved deformetrica package importing. Example: `import deformetrica as dfca`
- Bugfix: GUI toggle button values
- Updates:
    - pytorch to 1.1.0
    - pykeops to 1.1.1

## [4.2.0] - 2019-04-18
>>>>>>> dev
- Use a more efficient `_squared_distances` method in `AbstractKernel`. This highly increases performance. contributes to #39
- Improve TorchKernel
- Upgrades: Closes #45
    - pytorch to 1.0.1
    - pykeops to 1.0.1
    - numpy to 1.16
    - nibabel to 2.3.3
    - matplotlib to 3.0
    - pillow to 5.4
    - psutil to 5.4
- Better multiprocessing. high increases in performance. contributes to #39
- Multi-GPU support. contributes to #39
- Better OMP_NUM_THREADS management
- Fix image intensity normalization
- Add a check when working with meshes, to detect null area. fixes #41
- Add set_seed() method to the API
- Bugfix: Multiprocessing pool cleanup
- Python 3.7 compliant. Closes #40
- add `dtype` option to `model_options` in api and xml configuration file
- re-activate keops kernel on macos
- Bugfix: problem with image deformation
- Smarter kernel factory. Closes #37
- Better logger with console and file output. Closes #54
- Functional tests now pass with 1e-10 precision
- Added the longitudinal atlas model, along with functional tests. Closes #51.
- Added the MCMC-SAEM estimator, along with its own save-state system, and corresponding functional tests. Closes #19, #32.
- Refactoring of the deformable object class, now more robust.
- New VTK reader, based on the conda-available VTK library (v8). Closes #50.
- Generalized and harmonized use of the logger, along with automatic txt file dump. Closes #54. 
- Renamed `number-of-threads` to `number-of-processes` to better reflect reality
- Replace `use_cuda` boolean with `gpu_mode` enum. Available gpu_modes are: auto, full, none, kernel

## [4.1.0] - 2018-09-14
### Added
- New API: deformetrica can now be instantiated from python
- Automatic dimension detection from input file
- Allow to use different kernel types for the "shoot" and "flow" operations. Exponential class can now specify `shoot_kernel` and/or `flow_kernel`. resolves #10
- Allow device selection when instantiation a kernel.
  'device' kwarg is now available when using the kernel factory and 'device-device' is now available when using the xml configuration file. resolves #13
- New Python (PyQt5) GUI. Allows the configuration and run of a 'deterministic atlas'. This is an alpha-release
- Update Pykeops to version 0.0.14
- Bugfix: Gradients not computed when number_of_processes>1 and tensor_scalar_type is a FloatTensor. resolves #27
- Bugfix: Memory leak when using estimate_longitudinal_atlas due to pytorch's autograd graph. resolves #33

### Changed
- Split CLI commands into 2 groups: estimate and compute. resolves #6


## [4.0.2] - 2018-08-23
- Corrects the sobolev gradient in the multi-object case


## [4.0.1] - 2018-06-27
- Add different polyline VTK format


## [4.0.0] - 2018-06-14
### Added
- Bugfix: version file not found. issue #24
- Easy install with `conda install -c pytorch -c conda-forge -c anaconda -c aramislab deformetrica`, without any manual compilation.
- All existing deformetrica functionalities now work with 2d or 3d gray level images.
- A L-BFGS optimization method can now be used for registration, regression, deterministic and bayesian atlases.
- Gradients are now automagically computed using PyTorch's autograd.
- It is now possible to perform all computations on the gpu through the `use-cuda` option.

### Changed
- C++ is replaced by Python.
- The "exact" kernel is now named "torch"; the "cudaexact" kernel is now named "keops".
- The "deformable-object-type" xml entry is now split in two entries: "deformable-object-type" and "attachment-type". With this renamming, "NonOrientedSurfaceMesh" becomes a "SurfaceMesh" with a "Varifold" attachment (and an "OrientedSurfaceMesh" a "SurfaceMesh" with a "Current" attachment).

### Removed
- The Nesterov scheme for the gradient ascent optimizer (which was named "FastGradientAscent") is not available anymore. L-BFGS is more efficient though!


benchmark/parallelization.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-
import gc

import pykeops
from api import Deformetrica
import support.utilities as utilities
from unit_tests import example_data_dir, sandbox_data_dir
import os
import time
import torch


data_dir = os.path.join(os.path.dirname(__file__), "data")

# # SMALL 3D MESH
# dataset_specifications = {
#     'dataset_filenames': [
#         [{'amygdala': example_data_dir + '/atlas/landmark/3d/brain_structures/data/amygdala1.vtk',
#           'hippo': example_data_dir + '/atlas/landmark/3d/brain_structures/data/hippo1.vtk'}],
#         # [{'amygdala': example_data_dir + '/atlas/landmark/3d/brain_structures/data/amygdala2.vtk',
#         #   'hippo': example_data_dir + '/atlas/landmark/3d/brain_structures/data/hippo2.vtk'}],
#         # [{'amygdala': example_data_dir + '/atlas/landmark/3d/brain_structures/data/amygdala3.vtk',
#         #   'hippo': example_data_dir + '/atlas/landmark/3d/brain_structures/data/hippo3.vtk'}],
#         # [{'amygdala': example_data_dir + '/atlas/landmark/3d/brain_structures/data/amygdala4.vtk',
#         #   'hippo': example_data_dir + '/atlas/landmark/3d/brain_structures/data/hippo4.vtk'}]
#         ],
#     # 'subject_ids': ['subj1', 'subj2', 'subj3', 'subj4']
#     'subject_ids': ['subj1']
# }
# template_specifications = {
#     'amygdala': {'deformable_object_type': 'SurfaceMesh',
#                  'kernel_type': 'keops', 'kernel_width': 15.0,
#                  'noise_std': 10.0,
#                  'filename': example_data_dir + '/atlas/landmark/3d/brain_structures/data/amyg_prototype.vtk',
#                  'attachment_type': 'varifold'},
#     'hippo': {'deformable_object_type': 'SurfaceMesh',
#               'kernel_type': 'keops', 'kernel_width': 15.0,
#               'noise_std': 6.0,
#               'filename': example_data_dir + '/atlas/landmark/3d/brain_structures/data/hippo_prototype.vtk',
#               'attachment_type': 'varifold'}
# }

# LARGE 3D mesh
dataset_specifications = {'dataset_filenames': [], 'subject_ids': []}
for file in os.listdir(data_dir + '/landmark/3d/right_hippocampus_2738'):
    subject_id, visit_age = utilities.adni_extract_from_file_name(file)

    dataset_specifications['dataset_filenames'].append(
        [{'hippo': data_dir + '/landmark/3d/right_hippocampus_2738/' + file}],
    )
    dataset_specifications['subject_ids'].append(subject_id)

template_specifications = {
    'hippo': {'deformable_object_type': 'SurfaceMesh',
              'kernel_type': 'keops', 'kernel_width': 15.0,
              'noise_std': 6.0,
              'filename': data_dir + '/landmark/3d/right_hippocampus_2738/sub-ADNI002S0729_ses-M00.vtk',
              'attachment_type': 'varifold'}
}

# # FULL T1 IMAGE
# dataset_specifications = {
#     'dataset_filenames': [
#         # [{'brain': sandbox_data_dir + '/registration/image/3d/brains/data/s0021_7260_0.nii'}],
#         # [{'brain': sandbox_data_dir + '/registration/image/3d/brains/data/s0021_7960_8.nii'}],
#         [{'brain': sandbox_data_dir + '/registration/image/3d/brains/data/s0041_7110_0.nii'}]
#     ],
#     'subject_ids': ['s0021']
# }
# template_specifications = {
#     'brain': {'deformable_object_type': 'Image',
#               'kernel_type': 'keops', 'kernel_width': 10.0,
#               'noise_std': 1.0,
#               'filename': sandbox_data_dir + '/registration/image/3d/brains/data/s0021_7260_0.nii',
#               'attachment_type': 'varifold'}
# }


current_log_likelihood = 0.


def __estimator_callback(status_dict):
    global current_log_likelihood
    # current_iteration = status_dict['current_iteration']
    current_log_likelihood = status_dict['current_log_likelihood']
    return True


def deterministic_atlas_3d_brain_structure(kernel_type, nb_process, process_per_gpu, kernel_width):

    downsampling_factor = max(1, int(kernel_width / 2))
    logger.info('downsampling_factor=' + str(downsampling_factor))

    template_specifications['hippo']['kernel_type'] = kernel_type
    template_specifications['hippo']['kernel_width'] = kernel_width

    with Deformetrica(verbosity='DEBUG') as deformetrica:
        deformetrica.estimate_deterministic_atlas(
            template_specifications,
            dataset_specifications,
            estimator_options={'optimization_method_type': 'GradientAscent', 'max_iterations': 5,
                               'use_cuda': True, 'callback': __estimator_callback},
            model_options={'deformation_kernel_type': kernel_type, 'deformation_kernel_width': kernel_width, 'deformation_kernel_device': 'cuda',
                           'downsampling_factor': downsampling_factor,
                           'number_of_processes': nb_process, 'process_per_gpu': process_per_gpu},
            write_output=False)


def registration_3d_image(nb_process, number_of_time_points, kernel_width):

    downsampling_factor = max(1, int(kernel_width/2))
    logger.info('downsampling_factor=' + str(downsampling_factor))

    template_specifications['brain']['kernel_width'] = kernel_width

    with Deformetrica(verbosity='DEBUG') as deformetrica:
        deformetrica.estimate_registration(template_specifications, dataset_specifications,
                                           estimator_options={'optimization_method_type': 'GradientAscent', 'max_iterations': 20,
                                                              'use_cuda': False, 'callback': __estimator_callback},
                                           model_options={'deformation_kernel_type': 'keops', 'deformation_kernel_width': kernel_width,
                                                          'number_of_time_points': number_of_time_points, 'downsampling_factor': downsampling_factor,
                                                          'number_of_processes': nb_process, 'process_per_gpu': 1},
                                           write_output=False)


RUN_CONFIG = [
    # kernel_type, nb_process, process_per_gpu, kernel_width
    # (deterministic_atlas_3d_brain_structure, 'keops', 12, 1, 10.0),
    # (deterministic_atlas_3d_brain_structure, 'keops', 24, 1, 10.0),
    # (deterministic_atlas_3d_brain_structure, 'keops', 1, 1, 10.0),

    (deterministic_atlas_3d_brain_structure, 'keops', 2, 1, 10.0),
    (deterministic_atlas_3d_brain_structure, 'keops', 3, 1, 10.0),
    (deterministic_atlas_3d_brain_structure, 'keops', 4, 1, 10.0),
    (deterministic_atlas_3d_brain_structure, 'keops', 6, 1, 10.0),
    (deterministic_atlas_3d_brain_structure, 'keops', 8, 1, 10.0),
    (deterministic_atlas_3d_brain_structure, 'keops', 10, 1, 10.0),
    (deterministic_atlas_3d_brain_structure, 'keops', 12, 1, 10.0),
    (deterministic_atlas_3d_brain_structure, 'keops', 16, 1, 10.0),
    (deterministic_atlas_3d_brain_structure, 'keops', 20, 1, 10.0),
    (deterministic_atlas_3d_brain_structure, 'keops', 24, 1, 10.0),
    (deterministic_atlas_3d_brain_structure, 'keops', 28, 1, 10.0),
    (deterministic_atlas_3d_brain_structure, 'keops', 32, 1, 10.0),
    (deterministic_atlas_3d_brain_structure, 'keops', 36, 1, 10.0),

    # nb_process, number_of_time_points, kernel_width
    # (registration_3d_image, 1, 3, 10.0),
    # (registration_3d_image, 1, 5, 10.0),
    # (registration_3d_image, 1, 7, 10.0),
    # (registration_3d_image, 1, 9, 10.0),
    # (registration_3d_image, 1, 11, 10.0),
]


if __name__ == "__main__":
    logger.info('torch.__version__=' + torch.__version__)
    logger.info('pykeops.__version__=' + pykeops.__version__)

    res_elapsed_time = []
    res_log_likelihood = []

    for current_run_config in RUN_CONFIG:
        func, *args = current_run_config
        logger.info('>>>>>>>>>>>>> func=' + str(func) + ', args=' + str(args))

        start = time.perf_counter()
        func(*args)
        elapsed_time = time.perf_counter()-start
        logger.info('elapsed_time: ' + str(elapsed_time))
        logger.info('current_log_likelihood: ' + str(current_log_likelihood))

        res_elapsed_time.append(elapsed_time)
        res_log_likelihood.append(current_log_likelihood)

        # cleanup between runs
        time.sleep(0.5)
        if torch.cuda.is_available():
            torch.cuda.empty_cache()
        gc.collect()
        time.sleep(0.5)

    logger.info('===== RESULTS =====')
    logger.info(res_elapsed_time)
    logger.info(res_log_likelihood)

    # assert len(nb_processes) == len(results)
    #
    # index = np.arange(len(RUN_CONFIG))
    # bar_width = 0.2
    # opacity = 0.4
    #
    # fig, ax = plt.subplots()
    #
    # ax.bar(index + bar_width, results, bar_width, label=':')
    #
    # ax.set_xlabel('Nb process')
    # ax.set_ylabel('Runtime (s)')
    # ax.set_title('Runtime by number of processes')
    # # ax.set_xticks(index + bar_width * ((len(kernels)*len(initial_devices))/2) - bar_width/2)
    # # ax.set_xticklabels([r['setup']['tensor_size'] for r in results if r['setup']['device'] == 'cpu'])
    # ax.legend()
    #
    # fig.tight_layout()
    #
    # plt.show()
